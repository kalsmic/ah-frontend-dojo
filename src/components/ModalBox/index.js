// react libraries
import React from 'react';

// third-party libraries
import PropTypes from 'prop-types';

// styles
import 'components/ModalBox/ModalBox.scss';

// components
import { closeOpenModalFunction } from 'constants/staticsMethods';

/**
 * This is a function.
 * @param {*} n - mixed parameter types*
 * @return {component}
 * @example
 * foo = (props) => {}
 */
const ModalBox = (props) => {
  const {
    children, show, title, backdropId
  } = props;

  if (!show) {
    return null;
  }
  return (
    <div className="backdrop" id={backdropId}>
      <div className="modal">
        <header className="modal__header">
          <h4 className="modal__header__title">{title}</h4>
          <button
            className="modal__header__btn"
            onClick={function (e) {
              e.preventDefault();
              closeOpenModalFunction(backdropId);
            }}
            type="button"
          >
            <i>&times;</i>
          </button>
        </header>
        <div className="modal__section">
          {children}
        </div>

        <div className="modal__footer" />
      </div>
    </div>

  );
};

ModalBox.propTypes = {
  title: PropTypes.string.isRequired,
  backdropId: PropTypes.string.isRequired,
  show: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
};

export default ModalBox;
