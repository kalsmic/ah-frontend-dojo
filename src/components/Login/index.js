// React
import React, { Component } from 'react';

// third-party libraries
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

// login action
import loginUser from 'store/actions/loginActions';

// Import components
import InputBox from 'components/InputBox';
import Button from 'components/Button';

// Login form css
import 'components/Login/Login.scss';

// Modal Component
import ModalBox from 'components/ModalBox';
import { closeOpenModalFunction } from 'constants/staticsMethods';

export class Login extends Component {
  constructor(props) {
    super(props);


    this.state = {
      email: '',
      password: '',
      submitted: false,
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e) {
    const { name, value } = e.target;

    this.setState({ [name]: value });
  }

  handleSubmit(e) {
    e.preventDefault();

    this.setState({ submitted: true });

    const { email, password } = this.state;
    const { login } = this.props;

    if (email && password) {
      login(email, password);
    }
  }


  render() {
    const { loggingIn, error, success } = this.props;
    const { email, password, submitted } = this.state;
    return (
      <ModalBox title="LOGIN" show backdropId="loginModal">
        {success ? closeOpenModalFunction('loginModal') : ''}

        <div className="login">


          <form name="form" onSubmit={this.handleSubmit}>
            <div className={`login__input-group ${submitted && !email ? ' has-error' : ''}`}>
              <InputBox
                inputType="email"
                inputClass="login__input-control"
                name="email"
                value={email}
                handleChange={this.handleChange}
                placeholder="Email"
              >
                {submitted && !email
                && <div id="emailError" className="login__error">email is required</div>
                }
              </InputBox>

            </div>
            <div className={`login_input-group${submitted && !password ? ' has-error' : ''}`}>


              <InputBox
                inputType="password"
                inputClass="login__input-control"
                name="password"
                value={password}
                handleChange={this.handleChange}
                placeholder="Password"
              >
                {submitted && !password
                && <div id="passwordError" className="login__error">Password is required</div>
                }
              </InputBox>


            </div>
            {error && <div className="login__error">{error}</div>}

            <div className="input-group">
              <Button btnName="Login" btnClass="login__button" btnEvent={this.handleSubmit} />
              {loggingIn
              && (
                <img
                  alt="loading"
                  src="data:image/gif;base64,R0lGODlhEAAQAPIAAP///wAAAMLCwkJCQgAAAGJiYoKCgpKSkiH/C05FVFNDQVBFMi4wAwEAAAAh/hpDcmVhdGVkIHdpdGggYWpheGxvYWQuaW5mbwAh+QQJCgAAACwAAAAAEAAQAAADMwi63P4wyklrE2MIOggZnAdOmGYJRbExwroUmcG2LmDEwnHQLVsYOd2mBzkYDAdKa+dIAAAh+QQJCgAAACwAAAAAEAAQAAADNAi63P5OjCEgG4QMu7DmikRxQlFUYDEZIGBMRVsaqHwctXXf7WEYB4Ag1xjihkMZsiUkKhIAIfkECQoAAAAsAAAAABAAEAAAAzYIujIjK8pByJDMlFYvBoVjHA70GU7xSUJhmKtwHPAKzLO9HMaoKwJZ7Rf8AYPDDzKpZBqfvwQAIfkECQoAAAAsAAAAABAAEAAAAzMIumIlK8oyhpHsnFZfhYumCYUhDAQxRIdhHBGqRoKw0R8DYlJd8z0fMDgsGo/IpHI5TAAAIfkECQoAAAAsAAAAABAAEAAAAzIIunInK0rnZBTwGPNMgQwmdsNgXGJUlIWEuR5oWUIpz8pAEAMe6TwfwyYsGo/IpFKSAAAh+QQJCgAAACwAAAAAEAAQAAADMwi6IMKQORfjdOe82p4wGccc4CEuQradylesojEMBgsUc2G7sDX3lQGBMLAJibufbSlKAAAh+QQJCgAAACwAAAAAEAAQAAADMgi63P7wCRHZnFVdmgHu2nFwlWCI3WGc3TSWhUFGxTAUkGCbtgENBMJAEJsxgMLWzpEAACH5BAkKAAAALAAAAAAQABAAAAMyCLrc/jDKSatlQtScKdceCAjDII7HcQ4EMTCpyrCuUBjCYRgHVtqlAiB1YhiCnlsRkAAAOwAAAAAAAAAAAA=="
                />
              )
              }
              <i className="login__forgot__pwd">
                {' '}
                <Link to="/log" id="forgot">Forgot Password </Link>
                {' '}
                or signin with
              </i>
            </div>

            <div className="login__social">
              <a className="button--social-login button--facebook" href="fb">
                <i className="icon fa fa-facebook" />
              </a>
              <a className="button--social-login button--linkedin" href="#google">
                <i className="icon fa fa-google" />
              </a>

            </div>


          </form>
        </div>
      </ModalBox>
    );
  }
}

export const mapStateToProps = (state) => {
  const { loggingIn, success, error } = state.loginReducer;

  return {
    loggingIn,
    success,
    error,
  };
};

// Default props
Login.defaultProps = {
  loggingIn: false,
  error: '',
  success: false,
};

Login.propTypes = {
  login: PropTypes.func.isRequired,
  loggingIn: PropTypes.bool,
  error: PropTypes.string,
  success: PropTypes.bool,
};

export const mapDispatchToProps = dispatch => ({
  login(email, password) {
    dispatch(loginUser(email, password));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
