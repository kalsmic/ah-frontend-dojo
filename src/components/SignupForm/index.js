// react libraries
import React from 'react';

// third-party libraries
import PropTypes from 'prop-types';

// styles
import 'components/SignupForm/SignupForm.scss';

// components
import Button from 'components/Button';
import InputBox from 'components/InputBox';
import ModalBox from 'components/ModalBox';

//  string constants
import { SIGN_UP_STRING } from 'constants/constantStrings';

/**
/**
 * This is a function which return the SignupForm component
 * @param {*} n - mixed parameter types*
 * @return {component}
 * @example
 * foo = (props) => {}
 */
const SignupForm = (props) => {
  const {
    onChangeHandler, email, username, password,
    confirmPassword, onSubmitHandler, emailError, usernameError,
    passwordError, confirmPasswordError, signupMessageState, isSigningUp, backdropId,
  } = props;

  return (
    // Signup form enclosed inside the modal
    <ModalBox show title={SIGN_UP_STRING} {...props} backdropId={backdropId}>
      <div>
        <form className="signupForm" id="signupForm">
          <InputBox
            name="email"
            inputType="email"
            handleChange={onChangeHandler}
            placeholder="email"
            inputClass="signupForm__input"
            value={email}
          >
            <br />
            <span id="signupForm__input__error__email" className="signupForm__input__error">
              <small>{emailError}</small>
            </span>
          </InputBox>
          <InputBox
            name="username"
            inputType="text"
            handleChange={onChangeHandler}
            placeholder="username"
            inputClass="signupForm__input"
            value={username}
          >
            <br />
            <span id="signupForm__input__error__username" className="signupForm__input__error">
              <small>{usernameError}</small>
            </span>
          </InputBox>
          <InputBox
            name="password"
            inputType="password"
            handleChange={onChangeHandler}
            placeholder="password"
            inputClass="signupForm__input"
            value={password}
          >
            <br />
            <span id="signupForm__input__error__password" className="signupForm__input__error">
              <small>{passwordError}</small>
            </span>
          </InputBox>
          <InputBox
            name="confirmPassword"
            inputType="password"
            handleChange={onChangeHandler}
            placeholder="confirm password"
            inputClass="signupForm__input"
            value={confirmPassword}
          >
            <br />
            <span id="signupForm__input__error__confirmPassword" className="signupForm__input__error">
              <small>{confirmPasswordError}</small>
            </span>
          </InputBox>
          <br />
          <span className="signupForm__signUpMessage">{signupMessageState}</span>
          {isSigningUp && <div className="signupForm__loader" />}
          <Button
            id="signupForm__button"
            btnClass="signupForm__button"
            btnName={SIGN_UP_STRING}
            btnEvent={onSubmitHandler}
          />
        </form>
      </div>
    </ModalBox>
  );
};

// prop types validations
SignupForm.propTypes = {
  onChangeHandler: PropTypes.func,
  email: PropTypes.string,
  username: PropTypes.string,
  password: PropTypes.string,
  confirmPassword: PropTypes.string,
  onSubmitHandler: PropTypes.func,
  emailError: PropTypes.string,
  usernameError: PropTypes.string,
  passwordError: PropTypes.string,
  confirmPasswordError: PropTypes.string,
  signupMessageState: PropTypes.string,
  isSigningUp: PropTypes.bool,
  backdropId: PropTypes.string,
};

// default props
SignupForm.defaultProps = {
  onChangeHandler: () => { },
  email: '',
  username: '',
  password: '',
  confirmPassword: '',
  onSubmitHandler: () => { },
  emailError: '',
  usernameError: '',
  passwordError: '',
  confirmPasswordError: '',
  signupMessageState: '',
  isSigningUp: false,
  backdropId: 'signupModal',
};

export default SignupForm;
