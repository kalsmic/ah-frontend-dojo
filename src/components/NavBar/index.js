// react libraries
import React, { Component } from 'react';

// third party libraries
import { Link, Route } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

// styles
import './NavBarStyle.scss';


import SignupUserPage from 'components/SignupUser';
import { closeOpenModalFunction } from 'constants/staticsMethods';

import Login from 'components/Login';
import { logoutUser } from 'store/actions/loginActions';

export class Navbar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      opensignupModal: false,
      openloginModal: false,
    };
  }

  logout = (e) => {
    e.preventDefault();
    const { logout } = this.props;

    logout();
  };


  openModalHandler = (e) => {
    e.preventDefault();
    this.setState({ [`${e.target.id}`]: true });

    const openModalLink = document.getElementById(e.target.id);

    openModalLink.addEventListener('click', () => { e.preventDefault(); closeOpenModalFunction(e.target.id.slice(4)); });
  };

  render() {
    const { opensignupModal, openloginModal } = this.state;
    const { user } = this.props;


    return (
      <div className="navbar">
        {
          opensignupModal && (
            <Route
              render={
                props => (
                  <SignupUserPage
                    {...props}
                  />
                )
              }
            />
          )
        }
        {
          openloginModal && (
            <Route
              render={
                props => (
                  <Login
                    {...props}
                  />
                )
              }
            />
          )
        }


        <div className="navbar__branding">
          <div className="navbar__branding__name">
            <h1>
              {"Authors' Haven"}
            </h1>
          </div>
          <div className="navbar__branding__moto">
            <p>
              We provide blogging bliss and a
            </p>
            <p>
              reading experience second to none
            </p>
          </div>
        </div>
        <div className="navbar__navigation">
          <div className="navbar__navigation__auth" id="userInfo">

            {user && (
            <ul>

              <li className="navbar__navigation__auth__button">
                <span
                  className="navbar__navigation__auth__button--link"
                  style={{ cursor: 'pointer' }}
                >
                  { user.username.concat('|') }
                </span>
              </li>

              <li className="navbar__navigation__auth__button">
                <span
                  className="navbar__navigation__auth__button--link"
                  style={{ cursor: 'pointer' }}
                  onClick={this.logout}
                  role="button"
                  tabIndex="0"
                  onKeyPress={this.handleKeyPress}
                >
                  {' '}
                      Logout
                </span>
              </li>

            </ul>

            )
              }

            {!user && (

            <ul>
              <li className="navbar__navigation__auth__button">
                <span
                  id="openloginModal"
                  href="/"
                  className="navbar__navigation__auth__button--link"
                  style={{ cursor: 'pointer' }}
                  onClick={this.openModalHandler}
                  role="button"
                  tabIndex="0"
                  onKeyPress={this.handleKeyPress}
                >
                  Login
                </span>
              </li>

              {' '}
              <li className="navbar__navigation__auth__button">
                <span
                  id="opensignupModal"
                  href="/"
                  className="navbar__navigation__auth__button--link"
                  style={{ cursor: 'pointer' }}
                  onClick={this.openModalHandler}
                  role="button"
                  tabIndex="0"
                  onKeyPress={this.handleKeyPress}
                >
Sign Up
                </span>
              </li>
            </ul>
            )}

          </div>
          <div className="navbar__navigation__articles">
            <ul className="navbar__navigation__articles__ul">
              <li className="navbar__navigation__articles__ul__li">
                <Link className="navbar__navigation__articles__ul__li--button" to="/">
                  Home
                </Link>
              </li>
              <li className="navbar__navigation__articles__ul__li">
                <Link className="navbar__navigation__articles__ul__li--button" to="/articles">
                  Articles
                </Link>
              </li>
            </ul>
          </div>
        </div>
      </div>
    );
  }
}

Navbar.defaultProps = {
  user: {},
};

Navbar.propTypes = {
  logout: PropTypes.func.isRequired,
  user: PropTypes.shape({
    username: PropTypes.string,
    email: PropTypes.string,
    token: PropTypes.string,
  })
};

const mapStateToProps = (state) => {
  const { user } = state.loginReducer;
  return { user };
};

export const mapDispatchToProps = dispatch => ({
  logout() { (dispatch(logoutUser())); }
});

export default connect(mapStateToProps, mapDispatchToProps)(Navbar);
