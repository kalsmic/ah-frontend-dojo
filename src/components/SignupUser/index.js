// react libraries
import React, { Component } from 'react';

// third-party libraries
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

// components
import SignupForm from 'components/SignupForm';

// actions
import registerUserActions from 'store/actions/registerUserActions';

// constants
import { SIGNUP_SUCCESS_MESSAGE, CONFIRM_PASSWORD_ERROR } from 'constants/constantStrings';

export class SignupUser extends Component {
  /**
  * @type {*} n
  */
  static propTypes = {
    signUp: PropTypes.func.isRequired,
    signupSuccess: PropTypes.string,
    signupError: PropTypes.shape({
      errors: PropTypes.shape({
        username: PropTypes.string,
        email: PropTypes.string,
        password: PropTypes.string,
      }),
    }),
    isSigningUp: PropTypes.bool,
  };

  /**
  * @type {*} n
  */
  static defaultProps = {
    signupError: { errors: { username: '', email: '', password: '' } },
    signupSuccess: '',
    isSigningUp: false,
  };


  /**
   * This is a function where the conponent state is initialized
   * @@type {*} n
   *
   */
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      username: '',
      password: '',
      confirmPassword: '',
    };
  }

  /**
   * This is a function which handles onclick events for the
   * button on the signup form
   * @param {event} - An event param*
   * @example
   * on-submission handler for the form
   *     submit(event)
   * submit user details on click
   */
  onSubmitHandler = (event) => {
    event.preventDefault();
    const {
      email, username, password,
    } = this.state;
    const { signUp } = this.props;
    signUp(email, username, password);
  }

  /**
   * This is a function which handles onchange events for
   * the input fields for the signup form
   * @param {event} - An event param*
   * @example
   * onchange handler for the form fields
   *     foo(event)
   * change update state on input field value change
   */
  onChangeHandler = (event) => {
    const { name, value } = event.target;
    this.setState({
      [name]: value,
    });
  };

  /**
  * This is a function which returns error messages
  * @param {dictionary} - An dictionary param*
  * @return {string}
  * @example
  *      foo({error:{username: 'error usernam'}})
  * extract string values from the response payloads
  */
  convertToString = (errorMessage) => {
    let returnedString = '';
    if ((typeof errorMessage) !== 'undefined') {
      [returnedString] = errorMessage;
    }
    return returnedString;
  };

  /**
  * This is a function which sets the success message on signup
  * @param {string} - A string param*
  * @return {string}
  * @example
  *      foo({error:{username: 'error username'}})
  * check if signing up was successful
  */
  logSuccessSignup = (signupSuccess) => {
    let signupMessage = '';
    if (signupSuccess.length > 0) {
      signupMessage = SIGNUP_SUCCESS_MESSAGE;
    }
    return signupMessage;
  };

  /**
  * This is a function which checks for password and
  * confirm password similarity
  * @param {(string\|2)} - strings params*
  * @return {string}
  * @example
  *      convert({error:{username: 'error usernam'}})
  * validate confirm password field
  */
  checkForSamePassword = (password, confirmPassword) => {
    let errorMessage = '';
    if (password !== confirmPassword) {
      errorMessage = CONFIRM_PASSWORD_ERROR;
    }
    return errorMessage;
  };

  /**
   * This is a function which renders the SignupUser component
   * @param {none} - An event param*
   * @return {component}
   * @example
   *     render(<Componrnt />)
   */
  render() {
    const {
      email, username, password, confirmPassword,
    } = this.state;
    const {
      signupError, signupSuccess, isSigningUp,
    } = this.props;

    return (
      <div>
        <SignupForm
          email={email}
          username={username}
          password={password}
          confirmPassword={confirmPassword}
          onChangeHandler={this.onChangeHandler}
          onSubmitHandler={this.onSubmitHandler}
          emailError={this.convertToString(signupError.email)}
          usernameError={this.convertToString(signupError.username)}
          passwordError={this.convertToString(signupError.password)}
          confirmPasswordError={this.checkForSamePassword(password, confirmPassword)}
          signupMessageState={this.logSuccessSignup(signupSuccess)}
          closeModal={this.closeModelHandler}
          isSigningUp={isSigningUp}
          backdropId="signupModal"
        />
      </div>
    );
  }
}

/**
* This is a function which maps the props to state
* @param {dictionary} - A dictionary of state param*
* @return {dictionary}
* @example
*      foo({isSigningUp: true})
*/
const mapStateToProps = (state) => {
  const {
    signupSuccess, signupError, isSigningUp,
  } = state.registerReducer;
  return {
    signupSuccess, signupError, isSigningUp,
  };
};

/**
* This is a function shich dispatches signup function to props
* @param {function} - An in-built function param*
* @return {function}
* @example
*      foo(dispatch)
*/
const mapDispatchToProps = dispatch => ({
  signUp: (email, username, password) => {
    dispatch(registerUserActions(email, username, password));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(SignupUser);
