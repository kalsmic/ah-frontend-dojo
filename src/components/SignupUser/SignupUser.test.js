// react libraries
import React from 'react';

// third-party libraries
import { shallow } from 'enzyme';

// components
import { SignupUser } from 'components/SignupUser';
/**
* @type {*} n
*/
const props = {
  onSubmitHandler: jest.fn(),
  onChangeHandler: jest.fn(),
  registerUserActions: jest.fn(),
  signUp: jest.fn(),
  isSigningUp: true,
  email: '',
  username: '',
  password: '',
  backdropId: 'signupModal',
};

/**
 * This is a function which tests SignupUser component
 */
describe('SignupUser Component', () => {
  let wrapper;
  let instance;
  beforeEach(() => {
    wrapper = shallow(<SignupUser {...props} />);
    instance = wrapper.instance();
  });

  it('matches the snapshot', () => {
    const component = shallow(<SignupUser {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should handle signup error', () => {
    const event = {
      target: {
        type: 'submit',
        name: 'signup',
      },
    };
    const signupSuccess = instance.onChangeHandler(event);
    event.preventDefault = jest.fn();
    instance.onSubmitHandler(event);
    instance.onChangeHandler(event);
    expect(signupSuccess).toBe(undefined);
    instance.setState({ signupError: { errors: { username: 'username error', email: 'email error', password: 'password error' } } });
    expect(instance.props.isSigningUp).toBe(true);
  });
});
