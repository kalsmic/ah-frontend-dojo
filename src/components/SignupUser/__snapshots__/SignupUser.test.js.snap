// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupUser Component matches the snapshot 1`] = `
<div>
  <SignupForm
    backdropId="signupModal"
    confirmPassword=""
    confirmPasswordError=""
    email=""
    emailError=""
    isSigningUp={true}
    onChangeHandler={[Function]}
    onSubmitHandler={[Function]}
    password=""
    passwordError=""
    signupMessageState=""
    username=""
    usernameError=""
  />
</div>
`;
